Learning matches Nengo learning more closely

hunse opened the pull request:
This improves both the match between emulator and Loihi learning, and emulator and Nengo learning. The result is that everything matches Nengo better.

Based off of #137. Replaces #98.

TODO: I haven't tested this in extreme cases where we actually see overflow in the weights. I think some of these cases were referenced in #98. I (or someone) should look at these cases.

hunse commented:
Okay, I fixed the problem from #89 with overflowing (when initializing `test_multiple_pes` with `lambda x: 0`), and added a test for it. That doesn't actually test weight overflow; what we were encountering was other overflows (specifically in `q0`, the input accumulator). To get weights to overflow, I think the easiest way is to set the `wgtExp` for learning connections (which is always 4) down to e.g. 0. However, we currently don't have a way to set that locally for a specific test, so we couldn't add a unit test for that.

hunse commented:
I tried using the config system to set the weight exponent for learning connections. It wasn't pretty, and finally I gave up when core Nengo tests were failing because they added SPA-specific parameters which I then tried to copy but didn't exist in the new config objects because they weren't initialized. You can see the progress [here](https://github.com/nengo/nengo-loihi/tree/learn-nengo-config).

So now I've switched to just having an attribute on the Model to set this (see the new test). We can revisit the config stuff later if we wish. Fingers crossed everything passes now.

drasmuss commented on nengo_loihi/tests/test_learning.py:
```diff
@@ -64,6 +66,66 @@ def test_pes_comm_channel(allclose, plt, seed, Simulator, n_per_dim, dims):
 
 
+def test_pes_overflow(allclose, plt, seed, Simulator):
+    dims = 1
```
Would be good to test with `dims > 1` (so that we're checking this works correctly for multiple dimensions).

hunse replied:
done

hunse commented on nengo_loihi/loihi_interface.py:
```diff
@@ -168,7 +169,10 @@ def build_core(n2core, core):  # noqa: C901
             usesXepoch=1,
         )
-        n2core.stdpUcodeMem[0].word = 0x00102108  # 2^-7 learn rate
+        ucode = microcodegen_uci.ruleToUCode(['dw = u1*x1*y1*(2^-7)'],
```
done

hunse commented on nengo_loihi/builder.py:
```diff
@@ -657,18 +660,31 @@ def build_connection(model, conn):
-                assert isinstance(conn.learning_rule_type.pre_synapse,
+            rule_type = conn.learning_rule_type
+            if isinstance(rule_type, nengo.PES):
+                assert isinstance(rule_type.pre_synapse,
```
done

hunse commented on nengo_loihi/loihi_api.py:
```diff
@@ -22,6 +22,16 @@
+LEARN_FRAC = 7  # extra least-significant bits added to weights for learning
+
+
+def learn_overflow_bits(n_factors):
```
done

hunse commented on nengo_loihi/loihi_api.py:
```diff
@@ -106,6 +109,20 @@ def shift(x, s, **kwargs):
         return np.left_shift(x, s, **kwargs)
 
 
+def scale_pes_errors(error, scale=1.):
```
done

hunse commented on nengo_loihi/loihi_api.py:
```diff
@@ -106,6 +109,20 @@ def shift(x, s, **kwargs):
+    error = np.round(error).astype(np.int32)
+    q = error > 127
+    if np.any(q):
+        warnings.warn("PES error > 127, clipping (max %s)" % (error.max()))
```
done

hunse commented on nengo_loihi/loihi_api.py:
```diff
@@ -543,20 +566,34 @@ def validate(self, core=None):
+        Parameters
+        ----------
+        w : float ndarray
+            Weights to be discretized, in range -255 to 255.
```
done

hunse commented on nengo_loihi/loihi_cx.py:
```diff
@@ -303,9 +311,37 @@ def discretize(target, value):
+                if mag_int > 127:
+                    mag_int = 127
+                    mag_frac = 127
+                synapse.tracing_mag = mag_int + mag_frac / 128.
```
done

hunse commented on nengo_loihi/tests/test_learning.py:
```diff
@@ -29,34 +30,38 @@ def test_pes_comm_channel(allclose, plt, seed, Simulator, n_per_dim, dims):
+                    atol=0.1, rtol=0.05)
+    assert allclose(y_loihi[post_tmask], y_dpost[post_tmask],
+                    atol=0.1, rtol=0.05)
+    assert allclose(y_loihi, y_nengo, atol=0.15, rtol=0.1)
```
They fail before the next commit. I made this change to make testing easier as I developed the next commit.

hunse commented:
Ok, I think I've addressed all of Dan's comments.

drasmuss commented:
Everything looks good on the emulator, but `test_pes_comm_channel` is failing for me on hardware.

hunse commented:
Ok, I've fixed `test_pes_comm_channel` on hardware.

As part of that, I removed the fudge factor in the learning overflow. This had originally been to make the emulator more like the chip, but I don't think it actually did. It made the chip learn significantly more slowly than the emulator. With it removed, the chip learns slightly faster. I'm still not sure why this is; it's possible it has to do with how stochastic rounding happens on the chip. Anyway, the emulator and chip are closer now, so overall it's a win I think.

The two new commits can probably get squashed into existing ones if we want. I just kept them separate for now.

Finally, as part of our testing for this, we should probably look at the adaptive control notebook and check it on the emulator and the chip.

hunse commented on nengo_loihi/loihi_cx.py:
```diff
@@ -248,7 +252,7 @@ def discretize(target, value):
         w_max = max(w_maxs) if len(w_maxs) > 0 else 0
         b_max = np.abs(self.bias).max()
-        wgtExp = -7
+        wgtExp = 0
```
So this only affects cases when there are no weights, or the weights are all zero (i.e. if `w_max > 1e-8`, it goes into the first case in the `if` block below and this value gets ignored/overwritten).

I think the reason I changed it is that it was causing problems when learning on connections with all the weights set to zero. I can't remember exactly what the problems were, though.

drasmuss commented:
Confirmed that the tests pass and notebook looks good on hardware 👌

hunse commented:
Ok, I think I've tested almost all the new added lines.

There's one emulator error for spikes being dropped in the trace that is not being tested. I'm not quite sure how to test this: I think we'd need some sort of input that's spiking really fast (possibly just having a number of neurons converge on the same input axon in the post population would do it). I'd definitely want to do some sort of test that makes sure the emulator matches the chip in this respect, so I think for now I'd recommend just leaving this untested.

There's also a line in `discretize_weights` for `lossy_shift = False` that isn't tested. I mostly added this in to do quick comparisons of what things would be like if the chip didn't do this. The current chip does do this, though, so it's not tested without anywhere. Again, I'm inclined to say that's fine.
